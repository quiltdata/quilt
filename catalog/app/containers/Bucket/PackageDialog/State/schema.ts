import type { ErrorObject } from 'ajv'
import * as React from 'react'
import * as R from 'ramda'

import * as AWS from 'utils/AWS'
import {
  JsonSchema,
  makeSchemaDefaultsSetter,
  makeSchemaValidator,
} from 'utils/JSONSchema'
import * as Request from 'utils/useRequest'
import * as Types from 'utils/types'
import * as workflows from 'utils/workflows'

import { metadataSchema, objectSchema } from '../../requests'

export type SchemaStatus =
  | { _tag: 'idle' }
  | { _tag: 'loading' }
  | { _tag: 'error'; error: Error }
  | { _tag: 'ready'; schema?: JsonSchema }

export const Idle = { _tag: 'idle' as const }
export const Loading = { _tag: 'loading' as const }
export const Err = (error: Error) => ({ _tag: 'error' as const, error })
export const Ready = (schema?: JsonSchema) => ({ _tag: 'ready' as const, schema })

export function mkMetaValidator(schema?: JsonSchema) {
  const schemaValidator = makeSchemaValidator(schema)
  return function validateMeta(value: Types.Json): (ErrorObject | Error)[] | undefined {
    const jsonObjectErr = value && !R.is(Object, value)
    if (jsonObjectErr) {
      return [new Error('Metadata must be a valid JSON object')]
    }

    const setDefaults = makeSchemaDefaultsSetter(schema)
    const errors = schemaValidator(setDefaults(value))
    if (errors.length) return errors
  }
}

export function useMetadataSchema(workflow?: workflows.Workflow): SchemaStatus {
  const s3 = AWS.S3.use()
  const schemaUrl = workflow?.schema?.url
  const req = React.useCallback(() => metadataSchema({ s3, schemaUrl }), [schemaUrl, s3])
  const result = Request.use(req, !!schemaUrl)

  if (!schemaUrl) return Ready()

  if (result === Request.Idle) return Idle
  if (result === Request.Loading) return Loading
  if (result instanceof Error) return Err(result)

  return Ready(result)
}

export function useEntriesSchema(workflow?: workflows.Workflow): SchemaStatus {
  const s3 = AWS.S3.use()
  const schemaUrl = workflow?.entriesSchema || ''
  const req = React.useCallback(() => objectSchema({ s3, schemaUrl }), [schemaUrl, s3])
  const result = Request.use(req, !!schemaUrl)

  if (!schemaUrl) return Ready()

  if (result === Request.Idle) return Idle
  if (result === Request.Loading) return Loading
  if (result instanceof Error) return Err(result)

  return Ready(result)
}
