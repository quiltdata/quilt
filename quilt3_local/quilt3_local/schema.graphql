directive @admin on FIELD_DEFINITION

scalar Datetime

scalar Json

scalar JsonRecord

type ContentIndexingSettings {
  extensions: [String!]!
  bytesDefault: Int!
  bytesMin: Int!
  bytesMax: Int!
}

type Config {
  contentIndexingSettings: ContentIndexingSettings! @admin
}

type BucketConfig {
  name: String!
  title: String!
  iconUrl: String
  description: String
  linkedData: Json
  overviewUrl: String
  tags: [String!]
  relevanceScore: Int!
  lastIndexed: Datetime
  snsNotificationArn: String @admin
  scannerParallelShardsDepth: Int @admin
  skipMetaDataIndexing: Boolean @admin
  fileExtensionsToIndex: [String!] @admin
  indexContentBytes: Int @admin
  permissions: [RoleBucketPermission!]! @admin
}

type UnmanagedRole {
  id: ID!
  name: String!
  arn: String
}

type ManagedRole {
  id: ID!
  name: String!
  arn: String
  permissions: [RoleBucketPermission!]!
}

union Role = UnmanagedRole | ManagedRole

enum BucketPermissionLevel {
  READ
  READ_WRITE
}

type RoleBucketPermission {
  role: Role!
  bucket: BucketConfig!
  level: BucketPermissionLevel
}

type AccessCountForDate {
  date: Datetime!
  value: Int!
}

type AccessCounts {
  total: Int!
  counts: [AccessCountForDate!]!
}

type PackageDir {
  path: String!
  metadata: JsonRecord
  # 32-bit Int is too small to represent dir size which quite often exceeds several GB
  size: Float!
  children: [PackageEntry!]!
}

type PackageFile {
  path: String!
  metadata: JsonRecord
  # 32-bit Int is too small to represent file size which quite often exceeds several GB
  size: Float!
  physicalKey: String!
}

union PackageEntry = PackageFile | PackageDir

type PackageRevision {
  hash: String!
  modified: Datetime!
  message: String
  metadata: JsonRecord!
  userMeta: JsonRecord
  totalEntries: Int
  # 32-bit Int is too small to represent package size which quite often exceeds several GB
  totalBytes: Float
  dir(path: String!): PackageDir
  file(path: String!): PackageFile
  accessCounts(window: Int = 30): AccessCounts
}

type PackageRevisionList {
  total: Int!
  page(number: Int = 1, perPage: Int = 10): [PackageRevision!]!
}

type Package {
  bucket: String!
  name: String!
  modified: Datetime!
  revisions: PackageRevisionList!
  revision(hashOrTag: String = "latest"): PackageRevision
  accessCounts(window: Int = 30): AccessCounts
}

enum PackageListOrder {
  NAME
  MODIFIED
}

type PackageList {
  total: Int!
  page(
    number: Int = 1
    perPage: Int = 30
    order: PackageListOrder = NAME
  ): [Package!]!
}

type Query {
  config: Config!
  bucketConfigs: [BucketConfig!]!
  bucketConfig(name: String!): BucketConfig
  packages(bucket: String!, filter: String): PackageList
  package(bucket: String!, name: String!): Package
  roles: [Role!]! @admin
  role(id: ID!): Role @admin
}

input BucketAddInput {
  name: String!
  title: String!
  iconUrl: String
  description: String
  linkedData: Json
  overviewUrl: String
  tags: [String!]
  relevanceScore: Int
  snsNotificationArn: String
  scannerParallelShardsDepth: Int
  skipMetaDataIndexing: Boolean
  fileExtensionsToIndex: [String!]
  indexContentBytes: Int
  setVersioning: Boolean
  delayScan: Boolean
}

input BucketUpdateInput {
  title: String!
  iconUrl: String
  description: String
  linkedData: Json
  overviewUrl: String
  tags: [String!]
  relevanceScore: Int
  snsNotificationArn: String
  scannerParallelShardsDepth: Int
  skipMetaDataIndexing: Boolean
  fileExtensionsToIndex: [String!]
  indexContentBytes: Int
  setVersioning: Boolean
}

type BucketAddSuccess {
  bucketConfig: BucketConfig!
}

type BucketNotFound {
  _: Boolean
}

type BucketAlreadyAdded {
  _: Boolean
}

type BucketDoesNotExist {
  _: Boolean
}

type BucketFileExtensionsToIndexInvalid {
  _: Boolean
}

type BucketIndexContentBytesInvalid {
  _: Boolean
}

type SnsInvalid {
  _: Boolean
}

type NotificationConfigurationError {
  _: Boolean
}

type NotificationTopicNotFound {
  _: Boolean
}

type InsufficientPermissions {
  _: Boolean
}

type IndexingInProgress {
  _: Boolean
}

union BucketAddResult =
    BucketAddSuccess
  | BucketAlreadyAdded
  | BucketDoesNotExist
  | BucketFileExtensionsToIndexInvalid
  | BucketIndexContentBytesInvalid
  | InsufficientPermissions
  | NotificationConfigurationError
  | NotificationTopicNotFound
  | SnsInvalid

type BucketUpdateSuccess {
  bucketConfig: BucketConfig!
}

union BucketUpdateResult =
    BucketUpdateSuccess
  | BucketFileExtensionsToIndexInvalid
  | BucketIndexContentBytesInvalid
  | BucketNotFound
  | NotificationConfigurationError
  | NotificationTopicNotFound
  | SnsInvalid

type BucketRemoveSuccess {
  _: Boolean
}

union BucketRemoveResult =
    BucketRemoveSuccess
  | BucketNotFound
  | IndexingInProgress

input PermissionInput {
  bucket: String!
  level: BucketPermissionLevel
}

input UnmanagedRoleInput {
  name: String!
  arn: String!
}

input ManagedRoleInput {
  name: String!
  permissions: [PermissionInput!]!
}

type RoleNameReserved {
  _: Boolean
}

type RoleNameExists {
  _: Boolean
}

type RoleNameInvalid {
  _: Boolean
}

type BucketConfigDoesNotExist {
  name: String!
}

type RoleDoesNotExist {
  _: Boolean
}

type RoleIsManaged {
  _: Boolean
}

type RoleIsUnmanaged {
  _: Boolean
}

type RoleCreateSuccess {
  role: Role!
}

union RoleCreateResult =
    RoleCreateSuccess
  | RoleNameReserved
  | RoleNameExists
  | RoleNameInvalid
  | BucketConfigDoesNotExist

type RoleUpdateSuccess {
  role: Role!
}

union RoleUpdateResult =
    RoleUpdateSuccess
  | RoleNameReserved
  | RoleNameExists
  | RoleNameInvalid
  | RoleIsManaged
  | RoleIsUnmanaged
  | BucketConfigDoesNotExist

type RoleDeleteSuccess {
  _: Boolean
}

type RoleAssigned {
  _: Boolean
}

union RoleDeleteResult =
    RoleDeleteSuccess
  | RoleDoesNotExist
  | RoleNameReserved
  | RoleAssigned

type Mutation {
  bucketAdd(input: BucketAddInput!): BucketAddResult! @admin
  bucketUpdate(name: String!, input: BucketUpdateInput!): BucketUpdateResult!
    @admin
  bucketRemove(name: String!): BucketRemoveResult! @admin

  roleCreateManaged(input: ManagedRoleInput!): RoleCreateResult! @admin
  roleCreateUnmanaged(input: UnmanagedRoleInput!): RoleCreateResult! @admin
  roleUpdateManaged(id: ID!, input: ManagedRoleInput!): RoleUpdateResult! @admin
  roleUpdateUnmanaged(id: ID!, input: UnmanagedRoleInput!): RoleUpdateResult!
    @admin
  roleDelete(id: ID!): RoleDeleteResult! @admin
}
