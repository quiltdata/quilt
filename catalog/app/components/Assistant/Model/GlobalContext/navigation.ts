import * as Eff from 'effect'
import * as React from 'react'
import * as RR from 'react-router-dom'
import { Schema as S } from '@effect/schema'

import bucketRoutes from 'containers/Bucket/Routes'
import search from 'containers/Search/Route'
import * as ROUTES from 'constants/routes'
import * as Log from 'utils/Logging'
import * as Nav from 'utils/Navigation'
import * as XML from 'utils/XML'

import * as Content from '../Content'
import * as Context from '../Context'
import * as Tool from '../Tool'

const MODULE = 'GlobalContext/navigation'

// the routes are in the order of matching
// TODO: specify/describe all the *relevant* routes
const routeList = [
  Nav.makeRoute({
    name: 'home',
    path: ROUTES.home.path,
    exact: true,
    description: 'Home page',
    // searchParams: S.Struct({
    //   // XXX: passing this param doesn't actually work bc of how it's implemented in
    //   //      website/pages/Landing/Buckets/Buckets.js
    //   q: SearchParamLastOpt.annotations({
    //     title: 'bucket filter query',
    //     description: 'filter buckets in the bucket grid',
    //   }),
    // }),
  }),
  Nav.makeRoute({
    name: 'install',
    path: ROUTES.install.path,
    description: 'Installation page',
  }),
  search,
  Nav.makeRoute({
    name: 'activate',
    path: ROUTES.activate.path,
    description: 'TBD',
  }),
  //         <Route path={paths.signIn} exact>
  //           <AuthSignIn />
  //         </Route>
  //         <Route path="/login" exact>
  //           <RedirectTo path={urls.signIn()} />
  //         </Route>
  //
  //         <Route path={paths.signOut} exact>
  //           <AuthSignOut />
  //         </Route>
  //
  //         {(cfg.passwordAuth === true || cfg.ssoAuth === true) && (
  //           <Route path={paths.signUp} exact>
  //             <AuthSignUp />
  //           </Route>
  //         )}
  //         {!!cfg.passwordAuth && (
  //           <Route path={paths.passReset} exact>
  //             <AuthPassReset />
  //           </Route>
  //         )}
  //         {!!cfg.passwordAuth && (
  //           <Route path={paths.passChange} exact>
  //             <AuthPassChange />
  //           </Route>
  //         )}
  //
  //         <Route path={paths.code} exact>
  //           <AuthCode />
  //         </Route>
  //
  //         <Route path={paths.activationError} exact>
  //           <AuthActivationError />
  //         </Route>
  //
  //         {cfg.mode === 'OPEN' && (
  //           // XXX: show profile in all modes?
  //           <Route path={paths.profile} exact>
  //             <OpenProfile />
  //           </Route>
  //         )}
  //
  //         <Route path={paths.admin}>
  //           <Admin />
  //         </Route>
  //
  //         <Route path={paths.uriResolver}>
  //           <UriResolver />
  //         </Route>
  //
  ...bucketRoutes,
] as const

type KnownRoute = (typeof routeList)[number]
type KnownRouteMap = {
  [K in KnownRoute['name']]: Extract<KnownRoute, { name: K }>
}
export const routes = Object.fromEntries(
  routeList.map((r) => [r.name, r]),
) as KnownRouteMap

export const NavigableRouteSchema = S.Union(
  ...routeList.map((r) => r.navigableRouteSchema),
)

type NavigableRoute = typeof NavigableRouteSchema.Type

export const locationFromRoute = (route: NavigableRoute) =>
  // @ts-expect-error
  S.encode(routes[route.name].paramsSchema)(route.params)

type History = ReturnType<typeof RR.useHistory>

const WAIT_TIMEOUT = Eff.Duration.seconds(30)
const NAV_LAG = Eff.Duration.seconds(1)

const navigate = (
  route: NavigableRoute,
  history: History,
  markers: Eff.SubscriptionRef.SubscriptionRef<Record<string, boolean>>,
) =>
  Log.scoped({
    name: `${MODULE}.navigate`,
    enter: [`to: ${route.name}`, Log.br, 'params:', route.params],
  })(
    Eff.pipe(
      locationFromRoute(route),
      Eff.Effect.tap((loc) => Eff.Effect.log(`Navigating to location:`, Log.br, loc)),
      Eff.Effect.andThen((loc) => Eff.Effect.sync(() => history.push(loc))),
      Eff.Effect.andThen(() =>
        Eff.Effect.gen(function* () {
          const { waitForMarkers } = routes[route.name]
          if (!waitForMarkers.length) return
          yield* Eff.Effect.log(`Waiting for markers: ${waitForMarkers.join(', ')}`)
          yield* Eff.Effect.sleep(NAV_LAG)
          yield* Eff.pipe(
            markers.changes,
            Eff.Stream.timeoutFail(() => ({ _tag: 'timeout' as const }), WAIT_TIMEOUT),
            Eff.Stream.runForEachWhile((currentMarkers) =>
              Eff.Effect.succeed(!waitForMarkers.every((k) => currentMarkers[k])),
            ),
            Eff.Effect.andThen(() => Eff.Effect.log('Markers found')),
            Eff.Effect.catchTag('timeout', () =>
              Eff.Effect.log(
                `Timed out after ${Eff.Duration.format(
                  WAIT_TIMEOUT,
                )} while waiting for markers`,
              ),
            ),
          )
        }),
      ),
    ),
  )

export interface Match {
  descriptor: KnownRoute
  decoded: NavigableRoute | null
}

export const matchLocation = (loc: typeof Nav.Location.Type): Match | null =>
  Eff.pipe(
    Eff.Array.findFirst(routeList, (route) =>
      RR.matchPath(loc.pathname, {
        path: route.path,
        exact: route.exact,
        strict: route.strict,
      })
        ? Eff.Option.some(route)
        : Eff.Option.none(),
    ),
    Eff.Option.map((descriptor) => ({
      descriptor,
      decoded: Eff.pipe(
        loc,
        // @ts-expect-error
        S.decodeOption(descriptor.paramsSchema),
        Eff.Option.map((params) => ({ name: descriptor.name, params }) as NavigableRoute),
        Eff.Option.getOrNull,
      ),
    })),
    Eff.Option.getOrNull,
  )

export function useRouteContext() {
  const loc = RR.useLocation()

  const match = React.useMemo(
    () => matchLocation({ pathname: loc.pathname, search: loc.search, hash: '' }),
    [loc.pathname, loc.search],
  )

  const description = React.useMemo(() => {
    if (!match) return ''
    const params = match.decoded?.params
      ? XML.tag('parameters', {}, JSON.stringify(match.decoded.params, null, 2))
      : null
    return XML.tag(
      'route-info',
      {},
      `Name: "${match.descriptor.name}"`,
      XML.tag('description', {}, match.descriptor.description),
      params,
    )
  }, [match])

  const msg = React.useMemo(
    () =>
      XML.tag(
        'viewport',
        {},
        XML.tag('current-location', {}, JSON.stringify(loc, null, 2)),
        description,
        'Refer to "navigate" tool schema for navigable routes and their parameters.',
      ).toString(),
    [description, loc],
  )

  return msg
}

export const NavigateSchema = S.Struct({
  route: NavigableRouteSchema,
}).annotations({
  title: 'navigate the catalog',
  description: 'navigate to a provided route',
})

export function useNavigate() {
  const history = RR.useHistory()
  const markers = Context.useMarkersRef()

  return Tool.useMakeTool(
    NavigateSchema,
    ({ route }) =>
      Eff.pipe(
        navigate(route, history, markers),
        Eff.Effect.match({
          onSuccess: () =>
            Tool.succeed(Content.text(`Navigating to the '${route.name}' route.`)),
          onFailure: (e) =>
            Tool.fail(
              Content.text(`Failed to navigate to the '${route.name}' route: ${e}`),
            ),
        }),
        Eff.Effect.map(Eff.Option.some),
      ),
    [history, markers],
  )
}
