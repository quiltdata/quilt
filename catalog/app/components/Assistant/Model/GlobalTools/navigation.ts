import * as Eff from 'effect'
import * as RR from 'react-router-dom'
import { Schema as S } from '@effect/schema'

import * as ROUTES from 'constants/routes'
import * as Log from 'utils/Logging'
import * as Nav from 'utils/Navigation'

import * as Content from '../Content'
import * as Tool from '../Tool'

const MODULE = 'GlobalTools/navigation'

const home = Nav.makeRoute({
  name: 'home',
  path: ROUTES.home.path,
  description: 'Home page',
  // searchParams: S.Struct({
  //   // XXX: passing this param doesn't actually work bc of how it's implemented in
  //   //      website/pages/Landing/Buckets/Buckets.js
  //   q: SearchParamLastOpt.annotations({
  //     title: 'bucket filter query',
  //     description: 'filter buckets in the bucket grid',
  //   }),
  // }),
})

const search = Nav.makeRoute({
  name: 'search' as const,
  path: ROUTES.home.path,
  description: 'Search page',
  // TODO: add search params
  // pathParams: emptyPathParams,
  // searchParams: emptySearchParams,
})

const routeList = [home, search] as const
type KnownRoute = (typeof routeList)[number]
type KnownRouteMap = {
  [K in KnownRoute['name']]: Extract<KnownRoute, { name: K }>
}
const routes = Object.fromEntries(routeList.map((r) => [r.name, r])) as KnownRouteMap

const NavigableRouteSchema = S.Union(...routeList.map((r) => r.navigableRouteSchema))

type NavigableRoute = typeof NavigableRouteSchema.Type

const NavigateSchema = S.Struct({
  route: NavigableRouteSchema,
}).annotations({
  title: 'navigate the catalog',
  description: 'navigate to a provided route',
})

type History = ReturnType<typeof RR.useHistory>

const navigate = (route: NavigableRoute, history: History) =>
  Log.scoped({
    name: `${MODULE}.navigate`,
    enter: [`to: ${route.name}`, Log.br, 'params:', route.params],
  })(
    Eff.pipe(
      route.params,
      S.encode(routes[route.name].paramsSchema),
      Eff.Effect.andThen((loc) => Eff.Effect.sync(() => history.push(loc))),
      Eff.Effect.match({
        onSuccess: () =>
          Tool.succeed(Content.text(`Navigating to the '${route.name}' route.`)),
        onFailure: (e) =>
          Tool.fail(
            Content.text(`Failed to navigate to the '${route.name}' route: ${e}`),
          ),
      }),
      Eff.Effect.map(Eff.Option.some),
    ),
  )

export function useNavigate() {
  const history = RR.useHistory()
  return Tool.useMakeTool(NavigateSchema, ({ route }) => navigate(route, history), [
    history,
  ])
}
