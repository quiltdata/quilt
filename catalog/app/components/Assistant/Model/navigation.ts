import * as Eff from 'effect'
import * as RR from 'react-router-dom'
import { Schema as S } from '@effect/schema'

import search from 'containers/Search/Route'
import * as ROUTES from 'constants/routes'
import * as Log from 'utils/Logging'
import * as Nav from 'utils/Navigation'

const MODULE = 'Assistant/Model/navigation'

const home = Nav.makeRoute({
  name: 'home',
  path: ROUTES.home.path,
  description: 'Home page',
  // searchParams: S.Struct({
  //   // XXX: passing this param doesn't actually work bc of how it's implemented in
  //   //      website/pages/Landing/Buckets/Buckets.js
  //   q: SearchParamLastOpt.annotations({
  //     title: 'bucket filter query',
  //     description: 'filter buckets in the bucket grid',
  //   }),
  // }),
})

const routeList = [home, search] as const
type KnownRoute = (typeof routeList)[number]
type KnownRouteMap = {
  [K in KnownRoute['name']]: Extract<KnownRoute, { name: K }>
}
const routes = Object.fromEntries(routeList.map((r) => [r.name, r])) as KnownRouteMap

export const NavigableRouteSchema = S.Union(
  ...routeList.map((r) => r.navigableRouteSchema),
)

export type NavigableRoute = typeof NavigableRouteSchema.Type

export type History = ReturnType<typeof RR.useHistory>

export const navigate = (route: NavigableRoute, history: History) =>
  Log.scoped({
    name: `${MODULE}.navigate`,
    enter: [`to: ${route.name}`, Log.br, 'params:', route.params],
  })(
    Eff.pipe(
      route.params,
      // @ts-expect-error
      S.encode(routes[route.name].paramsSchema),
      Eff.Effect.tap((loc) => Eff.Effect.log(`Navigating to location:`, Log.br, loc)),
      Eff.Effect.andThen((loc) => Eff.Effect.sync(() => history.push(loc))),
    ),
  )
