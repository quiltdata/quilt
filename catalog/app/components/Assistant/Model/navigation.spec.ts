import { Schema } from '@effect/schema'

import * as Nav from './navigation'

jest.mock(
  'constants/config',
  jest.fn(() => ({})),
)

describe('components/Assistant/Model/navigation', () => {
  describe('NavigableRouteSchema', () => {
    it('shoult decode input', () => {
      const routeInput = {
        name: 'search',
        params: {
          params: {
            resultType: 'o',
          },
        },
      }
      const routeDecoded = Schema.decodeUnknownSync(Nav.NavigableRouteSchema)(routeInput)
      expect(routeDecoded).toMatchSnapshot()
    })
  })
})
