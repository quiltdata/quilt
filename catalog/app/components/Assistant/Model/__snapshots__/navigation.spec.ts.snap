// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Assistant/Model/navigation NavigableRouteSchema shoult decode input 1`] = `
{
  "name": "search",
  "params": {
    "params": {
      "resultType": "o",
    },
  },
}
`;
